cypress/e2e/wallet-connection/switch-network.test.ts

```diff
@@ -0,0 +1,124 @@
+import { createDeferredPromise } from '../../../src/test-utils/promise'
+import { getTestSelector } from '../../utils'
+
+function getChainSelector(activeChain: string) {
+  cy.get(getTestSelector('chain-selector-logo')).invoke('attr', 'alt').should('eq', activeChain)
+  return cy.get(getTestSelector('chain-selector')).eq(1)
+}
+
+describe('network switching', () => {
+  beforeEach(() => {
+    cy.visit('/swap', { ethereum: 'hardhat' })
+    cy.get(getTestSelector('web3-status-connected'))
+  })
+
+  function rejectsNetworkSwitchWith(rejection: unknown) {
+    cy.hardhat().then((hardhat) => {
+      // Reject network switch
+      const switchChainStub = cy.stub(hardhat.provider, 'send').log(false).as('switch')
+      switchChainStub.withArgs('wallet_switchEthereumChain').rejects(rejection)
+      switchChainStub.callThrough() // allows other calls to return non-stubbed values
+    })
+
+    // Switch network
+    getChainSelector('Ethereum').click()
+    cy.contains('Polygon').click()
+
+    // Verify rejected network switch
+    cy.get('@switch').should('have.been.calledWith', 'wallet_switchEthereumChain')
+    getChainSelector('Ethereum')
+    cy.get(getTestSelector('web3-status-connected'))
+  }
+
+  it('should not display message on user rejection', () => {
+    const USER_REJECTION = { code: 4001 }
+    rejectsNetworkSwitchWith(USER_REJECTION)
+    cy.get(getTestSelector('popups')).should('not.contain', 'Failed to switch networks')
+  })
+
+  it('should display message on unknown error', () => {
+    rejectsNetworkSwitchWith(new Error('Unknown error'))
+    cy.get(getTestSelector('popups')).contains('Failed to switch networks')
+  })
+
+  it('should add missing chain', () => {
+    cy.hardhat().then((hardhat) => {
+      // https://docs.metamask.io/guide/rpc-api.html#unrestricted-methods
+      const CHAIN_NOT_ADDED = { code: 4902 } // missing message in useSelectChain
+
+      // Reject network switch with CHAIN_NOT_ADDED
+      const switchChainStub = cy.stub(hardhat.provider, 'send').log(false).as('switch')
+      let added = false
+      switchChainStub
+        .withArgs('wallet_switchEthereumChain')
+        .callsFake(() => (added ? Promise.resolve(null) : Promise.reject(CHAIN_NOT_ADDED)))
```

Comment:
I think these stubs should be named after the method being stubbed, not the `withArgs` provided to the stub